pkg/controller/cronjob/cronjob_controller2.go

```diff
@@ -0,0 +1,504 @@
+			// or if our jobs list is newer than our cj status after a relist, or if someone intentionally created
+			// a job that they wanted us to adopt.
+
+			// TODO: maybe handle the adoption case?  Concurrency/suspend rules will not apply in that case, obviously, since we can't
+			// stop users from creating jobs if they have permission.  It is assumed that if a
+			// user has permission to create a job within a namespace, then they have permission to make any cronJob
+			// in the same namespace "adopt" that job.  ReplicaSets and their Pods work the same way.
+			// TBS: how to update cj.Status.LastScheduleTime if the adopted job is newer than any we knew about?
+		} else if found && IsJobFinished(&j) {
+			_, status := getFinishedStatus(&j)
+			deleteFromActiveList(cj, j.ObjectMeta.UID)
+			recorder.Eventf(cj, v1.EventTypeNormal, "SawCompletedJob", "Saw completed job: %s, status: %v", j.Name, status)
+		}
+	}
+
+	// Remove any job reference from the active list if the corresponding job does not exist any more.
+	// Otherwise, the cronjob may be stuck in active mode forever even though there is no matching
+	// job running.
+	for _, j := range cj.Status.Active {
+		if found := childrenJobs[j.UID]; !found {
+			recorder.Eventf(cj, v1.EventTypeNormal, "MissingJob", "Active job went missing: %v", j.Name)
+			deleteFromActiveList(cj, j.UID)
+		}
+	}
+
+	updatedCJ, err := cjc.UpdateStatus(cj)
+	if err != nil {
+		klog.Errorf("Unable to update status for %s (rv = %s): %v", nameForLog, cj.ResourceVersion, err)
+		return err, nil
+	}
+	*cj = *updatedCJ
+
+	if cj.DeletionTimestamp != nil {
+		// The CronJob is being deleted.
+		// Don't do anything other than updating status.
+		return fmt.Errorf("cronjob %s/%s is being deleted", cj.Namespace, cj.Name), nil
+	}
+
+	if cj.Spec.Suspend != nil && *cj.Spec.Suspend {
+		klog.V(4).Infof("Not starting job for %s because it is suspended", nameForLog)
+		return nil, nil
+	}
+
+	times, err := getRecentUnmetScheduleTimes(*cj, now)
+	switch {
+	case err != nil && len(times) == 0:
+		// too many missed jobs, schedule the next one on time and return
+		recorder.Eventf(cj, v1.EventTypeWarning, "TooManyMissedTimes", "Too many missed times for the cronjob, will schedule the next one", err)
+		klog.Error("Too many missed times for the cronjob, will schedule the next one", nameForLog, err)
+		return nil, nil
+	case err != nil:
+		// return error
+		recorder.Eventf(cj, v1.EventTypeWarning, "FailedNeedsStart", "Cannot determine if job needs to be started: %v", err)
+		klog.Errorf("Cannot determine if %s needs to be started: %v", nameForLog, err)
+		return err, nil
+	case len(times) == 0:
+		// no unmet start time, return
+		klog.V(4).Infof("No unmet start times for %s", nameForLog)
+		return nil, nil
+	default:
+		// multiple unmet start times, start the last one.
+		klog.V(4).Infof("Multiple unmet start times for %s so only starting last one", nameForLog)
+	}
+
+	scheduledTime := times[len(times)-1]
```

Comment:
`scheduledTime` will be a time from the past, if I'm not mistaken, so I'm not sure that's what you want to use in the next two conditions, you'll just want to re-queue it for the next execution time, and `getRecentUnmetScheduleTimes` doesn't return that. It only returns past times. You'll need some kind of `getNextScheduleTime`.

Reply:
my bad, I misunderstood our last conversation, I'll revert that change as I had before, good catch, +1